File: example_02.cpp
// Use something like this to compile:
// g++ example_02.cpp -o example_02  -std=c++11 -O3

#include <iostream>
#include <random>
#include <list>
#include <fstream>

using namespace std;

const double WORLD_WIDTH = 2560.0;
const double WORLD_HEIGHT = 1440.0;

class Agent {
public:
    double vmax = 0.0;
    double x;
    double y;
    double dx = 0.0;
    double dy = 0.0;
    bool is_alive = true;
    Agent* target = NULL;
    int age = 0;
    int energy = 0;

    Agent() {
        x = WORLD_WIDTH * (rand() / (RAND_MAX + 1.0));
        y = WORLD_HEIGHT * (rand() / (RAND_MAX + 1.0));
    }

    void update(const list<Agent*>& food) {
        age++;

        // we can't move
        if (vmax == 0.0) { return; }

        // target is dead, don't chase it further
        if ((target != NULL) && (target->is_alive == false)) {
            target = NULL;
        }

        // eat the target if close enough
        if (target != NULL) {
            double squared_dist = pow((x - target->x), 2) + pow((y - target->y), 2);
            if (squared_dist < 400) {
                target->is_alive = false;
                energy++;
            }
        }

        // agent doesn't have a target, find a new one
        if (target == NULL) {
            double min_dist = 9999999;
            Agent* min_agent = NULL;

            for (const auto a: food) {
                if (a->is_alive == true) {
                    double squared_dist = pow((x - a->x), 2) + pow((y - a->y), 2);
                    if (squared_dist < min_dist) {
                        min_dist = squared_dist;
                        min_agent = a;
                    }
                }
            }
            if (min_dist < 100000) {
                target = min_agent;
            }
        }

        // initialize forces to zero
        double fx = 0;
        double fy = 0;

        // move in the direction of the target, if any
        if (target != NULL) {
            fx += 0.1*(target->x - x);
            fy += 0.1*(target->y - y);
        }

        // update our direction based on the 'force'
        dx += 0.05 * fx;
        dy += 0.05 * fy;

        // slow down agent if it moves faster than its max velocity
        double velocity = sqrt(pow(dx, 2) + pow(dy, 2));
        if (velocity > vmax) {
            dx = (dx / velocity) * (vmax);
            dy = (dy / velocity) * (vmax);
        }

        // update position based on delta x/y
        x += dx;
        y += dy;

        // ensure it stays within the world boundaries
        x = max(x, 0.0);
        x = min(x, WORLD_WIDTH);
        y = max(y, 0.0);
        y = min(y, WORLD_HEIGHT);
    }
};


class Predator: public Agent {
    public:
    Predator() : Agent() {
         vmax = 2.5;
    }
};

class Prey: public Agent {
    public:
    Prey() : Agent() {
         vmax = 2.0;
    }
};

class Plant: public Agent {
    public:
    Plant() : Agent() {
         vmax = 0.0;
    }
};

int main(int argc, char *argv[]) {
    srand (time(NULL));

    std::ios_base::sync_with_stdio(false);
        
    list<Agent*> predators;
    list<Agent*> preys;
    list<Agent*> plants;

    list<Agent*> empty;

    // create initial agents
    for (int i = 0; i < 10; i++) {
        Predator *p = new Predator();
        predators.push_back(p);
    }
    for (int i = 0; i < 10; i++) {
        Prey *p = new Prey();
        preys.push_back(p);
    }
    for (int i = 0; i < 100; i++) {
        Plant *p = new Plant();
        plants.push_back(p);
    }

    std::ofstream outfile;
    outfile.open ("output.csv");

    int timestep = 0;
    outfile << timestep << ',' << "Title" << ',' << "Predator Prey Relationship / Example 02 / C++" << endl;

    while (timestep < 10000) {
        // update all agents
        for (auto p: predators) { p->update(preys); }
        for (auto p: preys) { p->update(plants); }
        //for (auto p: plants) { p->update(empty); }  // no need to update the plants; they do not move

        // handle eaten and create new plants
        plants.remove_if([](const Agent* a) { return (a->is_alive == false) ? true : false; });
        for (int i=0; i < 2; i++) { plants.push_back(new Plant()); };

        // handle eaten and create new preys
        preys.remove_if([](const Agent* a) { return (a->is_alive == false) ? true : false; });
        
        for (auto p: preys) {
            if (p->energy > 5) {
                p->energy = 0;
                Prey* np = new Prey();
                np->x = p->x + -20 + 40 * (rand() / (RAND_MAX + 1.0));
                np->y = p->y + -20 + 40 * (rand() / (RAND_MAX + 1.0));
                preys.push_back(np);
            }
        }

        // handle old and create new predators
        predators.remove_if([](const Agent* a) { return (a->age > 2000) ? true : false; });

        for (auto p: predators) {
            if (p->energy > 10) {
                p->energy = 0;
                Predator* np = new Predator();
                np->x = p->x + -20 + 40 * (rand() / (RAND_MAX + 1.0));
                np->y = p->y + -20 + 40 * (rand() / (RAND_MAX + 1.0));
                predators.push_back(np);
            }
        }

        // write data to output file
        /*
        for (const auto p: predators) {
            outfile << timestep << ',' << "Position" << ',' << "Predator" << ',' << p->x << ',' << p->y << endl;
        }

        for (const auto p: preys) {
            outfile << timestep << ',' << "Position" << ',' << "Prey" << ',' << p->x << ',' << p->y << endl;
        }

        for (const auto p: plants) {
            outfile << timestep << ',' << "Position" << ',' << "Plant" << ',' << p->x << ',' << p->y << endl;
        }
        */

        timestep++;
    }

    outfile.close();
    cout << predators.size() << ", " << preys.size() << ", " << plants.size() << endl;
}

File: example_02.py
import math
import random

from datetime import datetime
random.seed(datetime.now().timestamp())

WORLD_WIDTH = 2560
WORLD_HEIGHT = 1440

class Agent():
    def __init__(self, x=None, y=None):
        super().__init__()

        # default values
        self.vmax = 2.0

        # initial position
        self.x = x if x else random.randint(0, WORLD_WIDTH)
        self.y = y if y else random.randint(0, WORLD_HEIGHT)

        # initial velocity
        self.dx = 0
        self.dy = 0

        # inital values
        self.is_alive = True
        self.target = None
        self.age = 0
        self.energy = 0

    def update(self, food=()):
        self.age = self.age + 1

        # we can't move
        if self.vmax == 0:
            return

        # target is dead, don't chase it further
        if self.target and not self.target.is_alive:
            self.target = None

        # eat the target if close enough
        if self.target:
            squared_dist = (self.x - self.target.x) ** 2 + (self.y - self.target.y) ** 2
            if squared_dist < 400:
                self.target.is_alive = False
                self.energy = self.energy + 1

        # agent doesn't have a target, find a new one
        if not self.target:
            min_dist = 9999999
            min_agent = None
            for a in food:
                if a is not self and a.is_alive:
                    sq_dist = (self.x - a.x) ** 2 + (self.y - a.y) ** 2
                    if sq_dist < min_dist:
                        min_dist = sq_dist
                        min_agent = a
            if min_dist < 100000:
                self.target = min_agent

        # initalize 'forces' to zero
        fx = 0
        fy = 0

        # move in the direction of the target, if any
        if self.target:
            fx += 0.1*(self.target.x - self.x)
            fy += 0.1*(self.target.y - self.y)

        # update our direction based on the 'force'
        self.dx = self.dx + 0.05 * fx
        self.dy = self.dy + 0.05 * fy

        # slow down agent if it moves faster than it max velocity
        velocity = math.sqrt(self.dx ** 2 + self.dy ** 2)
        if velocity > self.vmax:
            self.dx = (self.dx / velocity) * (self.vmax)
            self.dy = (self.dy / velocity) * (self.vmax)

        # update position based on delta x/y
        self.x = self.x + self.dx
        self.y = self.y + self.dy

        # ensure it stays within the world boundaries
        self.x = max(self.x, 0)
        self.x = min(self.x, WORLD_WIDTH)
        self.y = max(self.y, 0)
        self.y = min(self.y, WORLD_HEIGHT)

class Predator(Agent):
    def __init__(self, x=None, y=None):
        super().__init__()
        self.vmax = 2.5

class Prey(Agent):
    def __init__(self, x=None, y=None):
        super().__init__()
        self.vmax = 2.0

class Plant(Agent):
    def __init__(self, x=None, y=None):
        super().__init__()
        self.vmax = 0


def main():
    # open the ouput file
    f = open('output.csv', 'w')
    print(0, ',', 'Title', ',', 'Predator Prey Relationship / Example 02 / Pthon', file=f)

    # create initial agents
    preys = [Prey() for i in range(10)]
    predators = [Predator() for i in range(10)]
    plants = [Plant() for i in range(100)]

    timestep = 0
    while timestep < 10000:
        # update all agents
        #[f.update() for f in plants]  # no need to update the plants; they do not move
        [a.update(food=plants) for a in preys]
        [a.update(food=preys) for a in predators]

        # handle eaten and create new plant
        plants = [p for p in plants if p.is_alive is True]
        plants = plants + [Plant() for i in range(2)]

        # handle eaten and create new preys
        preys = [p for p in preys if p.is_alive is True]

        for p in preys[:]:
            if p.energy > 5:
                p.energy = 0
                preys.append(Prey(x = p.x + random.randint(-20, 20), y = p.y + random.randint(-20, 20)))

        # handle old and create new predators
        predators = [p for p in predators if p.age < 2000]

        for p in predators[:]:
            if p.energy > 10:
                p.energy = 0
                predators.append(Predator(x = p.x + random.randint(-20, 20), y = p.y + random.randint(-20, 20)))

        # write data to output file
        #[print(timestep, ',', 'Position', ',', 'Predator', ',', a.x, ',', a.y, file=f) for a in predators]
        #[print(timestep, ',', 'Position',  ',', 'Prey', ',', a.x, ',', a.y, file=f) for a in preys]
        #[print(timestep, ',', 'Position',  ',', 'Plant', ',', a.x, ',', a.y, file=f) for a in plants]

        timestep = timestep + 1

    print(len(predators), len(preys), len(plants))

if __name__ == "__main__":
    main()
